src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1495,6 +1497,161 @@ static bool extendedTypeIsOrDerivedFrom(NamedTypeSymbol hiddenContainer, NamedTy
+        /// <summary>
+        /// We remove non-methods that are "hidden" (based on extended type) by more specific members.
+        /// But we collect all the methods, because we'll only know if they are "hidden" after we have the complete set of extension members.
+        ///
+        /// Once we have the complete set of extension members, we'll call <see cref="CompleteMergeExtensionLookupResultsHidingLessSpecific(LookupResult, ConsList{TypeSymbol}, ref CompoundUseSiteInfo{AssemblySymbol})"/>
+        /// It will remove methods that are "hidden" by any remaining non-methods.
+        /// It will also compact the set (deal with any null entries left in the result).
+        /// </summary>
+        private void MergeExtensionLookupResultsHidingLessSpecificNonMethods(LookupResult current, LookupResult additional, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (!current.IsMultiViable || !additional.IsMultiViable)
+            {
+                current.MergePrioritized(additional);
+                return;
+            }
+
+            var currentSymbols = current.Symbols;
+            var additionalSymbols = additional.Symbols;
+
+            var currentCount = currentSymbols.Count;
+            var additionalCount = additionalSymbols.Count;
+            Debug.Assert(currentCount > 0);
+            Debug.Assert(additionalCount > 0);
+
+            for (int i = 0; i < additionalCount; i++)
+            {
+                var additionalSymbol = additionalSymbols[i];
+
+                removeLessSpecificNonMethodCurrentSymbols(additionalSymbol, currentSymbols, out bool isAdditionalSymbolHidden,
+                    currentCount, basesBeingResolved, ref useSiteInfo);
+
+                if (!isAdditionalSymbolHidden)
+                {
+                    currentSymbols.Add(additionalSymbol);
+                }
+            }
+
+            return;
+
+            // Determine whether we want to add the additional symbol
+            // and potentially remove some "hidden" non-method current symbols in the process
+            void removeLessSpecificNonMethodCurrentSymbols(Symbol additionalSymbol, ArrayBuilder<Symbol> currentSymbols, out bool isAdditionalSymbolHidden,
+                int currentCount, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                isAdditionalSymbolHidden = false;
+                for (int j = 0; j < currentCount; j++)
+                {
+                    var currentSymbol = currentSymbols[j];
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    if (currentSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } currentContainerUnderlying
+                        || additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying)
```

Comment:
>additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying [](http://example.com/codeflow?start=27&length=105)

It looks like this can be taken out of the loop